### PR TITLE
gRPC - per-client interceptors

### LINKED ISSUE #21648: gRPC: per-client interceptors
### Description

For gRPC services we now have `@GlobalInterceptor`s and interceptors registered for specific servies.
We likely need something similar for clients, so that a user can attach an interceptor to a specific client

CC @cescoffier @mkouba 

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #21648
- make the RegisterInterceptor annotaion repeatable
- refactor server interceptors

### CODE REVIEW

**Comment by mkouba** [comment]
It was a little bit more complex than I thought but the PR is ready now ;-).

@michalszynkiewicz I've replaced the generated `InterceptorStorage` impl with a synthetic bean initialized through a recorder method. It should be a little bit easier to maintain...

IMPL NOTE: `@RegisterClientInterceptor` is currently implemented as a CDI qualifier, otherwise we wouldn't be able to use a shared `@Singleton` synthetic beans for the injected clients. However, this decision may be revisited in the future.

**Comment by mkouba** [comment]
Note to myself - add a test with multiple `@RegisterClientInterceptor` annotations declared on an injected client.

File: docs/src/main/asciidoc/grpc-service-consumption.adoc [hunk 1]
```diff
@@ -309,9 +318,28 @@ public class MyInterceptor implements ClientInterceptor {
...
+----
+import io.quarkus.grpc.GrpcClient;
+import io.quarkus.grpc.RegisterClientInterceptor;
+
+import hello.Greeter;
+
+@ApplicationScoped
+class MyBean {
+
+    @RegisterClinetInterceptor(MySpecialInterceptor.class) <1>
```
Comment (michalszynkiewicz) [inline docs/src/main/asciidoc/grpc-service-consumption.adoc:336]:
```suggestion
    @RegisterClientInterceptor(MySpecialInterceptor.class) <1>
```

File: docs/src/main/asciidoc/grpc-service-consumption.adoc [hunk 2]
```diff
@@ -295,10 +295,19 @@ public class MyService {
 
 == Client Interceptors
 
-You can implement a gRPC client interceptor by implementing an `@ApplicationScoped` bean implementing `io.grpc.ClientInterceptor`:
+A gRPC client interceptor can be implemented by a CDI bean that also implements the `io.grpc.ClientInterceptor` interface.
+If you want to apply the interceptor to any injected client then annotate the interceptor bean with `@io.quarkus.grpc.GlobalInterceptor`.
+You can also register the interceptor per injected client with `@io.quarkus.grpc.RegisterClientInterceptor`.
+This annotation is repetable.
```
Comment (cescoffier) [inline docs/src/main/asciidoc/grpc-service-consumption.adoc:301]:
```suggestion
This annotation is repeatable.
```

File: docs/src/main/asciidoc/grpc-service-consumption.adoc [hunk 3]
```diff
@@ -295,10 +295,19 @@ public class MyService {
 
 == Client Interceptors
 
-You can implement a gRPC client interceptor by implementing an `@ApplicationScoped` bean implementing `io.grpc.ClientInterceptor`:
+A gRPC client interceptor can be implemented by a CDI bean that also implements the `io.grpc.ClientInterceptor` interface.
+If you want to apply the interceptor to any injected client then annotate the interceptor bean with `@io.quarkus.grpc.GlobalInterceptor`.
+You can also register the interceptor per injected client with `@io.quarkus.grpc.RegisterClientInterceptor`.
```
Comment (cescoffier) [inline docs/src/main/asciidoc/grpc-service-consumption.adoc:300]: You should mention that this annotation is used on the client (on the injection site), while GlobalInterceptor is on the interceptor.

File: extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/interceptors/MySecondClientInterceptor.java [hunk 4]
```diff
@@ -28,19 +28,16 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
                         new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(responseListener) {
 
                             @Override
-                            protected Listener<RespT> delegate() {
-                                callTime = System.nanoTime();
-                                return super.delegate();
+                            public void onMessage(RespT message) {
+                                System.out.println("ONMESSAHE - #2");
```
Comment (cescoffier) [inline extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/interceptors/MySecondClientInterceptor.java:32]: Shoudl be removed.

File: extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/interceptors/MyFirstClientInterceptor.java [hunk 5]
```diff
@@ -28,19 +28,16 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
                         new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(responseListener) {
 
                             @Override
-                            protected Listener<RespT> delegate() {
-                                callTime = System.nanoTime();
-                                return super.delegate();
+                            public void onMessage(RespT message) {
+                                System.out.println("ONMESSAHE - #1");
```
Comment (cescoffier) [inline extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/interceptors/MyFirstClientInterceptor.java:32]: Should be removed.

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Review by cescoffier — APPROVED** [review]

**Review by michalszynkiewicz — APPROVED** [review]

**Pull request merged**